design/Extend-VolumePolicies-to-support-more-actions.md

```diff
@@ -0,0 +1,176 @@
+- This becomes cumbersome at scale.
+- Using `VolumePolicies`, the user can just specify 2 simple `VolumePolicies` like for csi supported volumes as `snapshot` action and rest can be backed up`fs-backup` action:
+```yaml
+version: v1
+volumePolicies:
+- conditions:
+    storageClass:
+    - gp2
+  action:
+    type: snapshot
+- conditions: {}
+  action:
+    type: fs-backup
+```
+
+**Use-case 2:**
+- A user wants to use `fs-backup` for nfs volumes pertaining to a particular server
+- In such a scenario the user can just specify a `VolumePolicy` like:
+```yaml
+version: v1
+volumePolicies:
+- conditions:
+    nfs:
+      server: 192.168.200.90
+  action:
+    type: fs-backup
+```
+## High-Level Design
+- When the VolumePolicy action is set as `fs-backup` the backup workflow modifications would be:
+  - We call [backupItem() -> backupItemInternal()](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/item_backupper.go#L95) on all the items that are to be backed up
+  - Here when we encounter [Pod as an item ](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/item_backupper.go#L195)
+  - We will have to modify the backup workflow to account for the `fs-backup` VolumePolicy action
+
+
+- When the VolumePolicy action is set as `snapshot` the backup workflow modifications would be:
+  - Once again, We call [backupItem() -> backupItemInternal()](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/item_backupper.go#L95) on all the items that are to be backed up
+  - Here when we encounter [Persistent Volume as an item](https://github.com/vmware-tanzu/velero/blob/d4128542590470b204a642ee43311921c11db880/pkg/backup/item_backupper.go#L253)
+  - And we call the [takePVSnapshot func](https://github.com/vmware-tanzu/velero/blob/d4128542590470b204a642ee43311921c11db880/pkg/backup/item_backupper.go#L508)
+  - We need to modify the takePVSnapshot function to account for the `snapshot` VolumePolicy action.
+
+
+## Detailed Design
+- Update VolumePolicy action type validation to account for `fs-backup` and `snapshot` as valid VolumePolicy actions.
+- Modifications needed for `fs-backup` action:
+  - Now based on the specification of volume policy on backup request we will decide whether to go via legacy pod annotations approach or the newer volume policy based fs-backup action approach.
+  - If there is a presence of volume policy(fs-backup/snapshot)  on the backup request that matches as an action for a volume we use the newer volume policy approach to get the list of the volumes for `fs-backup` action
+  - Else continue with the annotation based legacy approach workflow.
+
+- Modifications needed for `snapshot` action:
+  - In the [takePVSnapshot function](https://github.com/vmware-tanzu/velero/blob/d4128542590470b204a642ee43311921c11db880/pkg/backup/item_backupper.go#L508) we will check the PV fits the volume policy criteria and see if the associated action is `snapshot`
```

Comment:
For CSI snapshot, it is taken by CSI plugin through BIA on the Persistent Volume Claim.
At present, the CSI plugin decides whether to process a volume through CSI snapshot method, Velero workflow doesn't know anything about how the decision is made. 
Then how to apply volumePolicies to CSI snapshots? It is not ideal if we have to have CSI plugin access volumePolicies, right?

Please clarify the solution in this design.

Reply:
`csi-snapshots` and `csi-datamover` are not targeted in the scope of this design @Lyndon-Li

Reply:
So both the design and implementation don't support CSI snapshot and data mover?
Personally, I think at least the design (including the detailed design which demonstrates the feasibility of the implementation) should support all backup types. Considering the fact that fs-backup is gradually replaced by data mover and native snapshot is gradually replaced by CSI snapshot, if we only support fs-backup and native snapshot, I am afraid the design will lose lots of its value.

Reply:
@Lyndon-Li to be clear, we don't only support fs-backup and native snapshot, we support "fs-backup" and "snapshot", but "snapshot" will be either native snapshot or csi, following the same decision logic used with the legacy opt-in/opt-out approach when snapshot is chosen -- could be either csi or native snapshot. However, I discussed this with @shubham-pampattiwar and it does make sense to add some more specific actions for future implementation, but I think we should hold off on implementing those until the legacy opt-in/opt-out approach has been deprecated, otherwise things will be really confusing with the multiple ways of specifying things.

Reply:
@Lyndon-Li I have updated the design sections (non-goals and snapshot note). Also added future enhancement section, PTAL.